### PR TITLE
Support for workload identity in AGIC Helm installation

### PR DESCRIPTION
**Pull request by akshaysngupta**
## Checklist
- [x] The title of the PR is clear and informative
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] Issues addressed by the PR are mentioned in the description followed by `Fixes`.

## Description

This PR adds support for workload identity in AGIC when installed via Helm. AGIC will use to communicate with ARM to update Application Gateway.
Feature Documentation: https://learn.microsoft.com/en-us/azure/aks/workload-identity-overview

* updates the authorization code to use the latest `DefaultAzureCredentials` token provider which supports using workload identity.
* To use workload identity, type `workloadIdentity` will apply the required labels and annotations on Deployment and Service Account.
```yaml
armAuth:
  type: workloadIdentity
  identityClientID:  <>
```

## Testing
* Label and Annotation on Service Account
image
*  Label on Pod
image
* Pod has required environment variables to enable workload identity
image
* AGIC successfully updating AppGW
image

## Fixes

Fixes https://github.com/Azure/application-gateway-kubernetes-ingress/issues/1492

### CODE REVIEW

**Comment by jbpaux** [comment]
I would have thought more about something like that for workload identity since workload identity is not aadPodIdentity:
```
armAuth:
  type: workloadIdentity
  identityResourceID: <>
  identityClientID:  <>
```

**Comment by akshaysngupta** [comment]
@jbpaux good suggestion, will make the change.

**Comment by jbpaux** [comment]
Maybe you can also update the comment since it's not exactly that now:
https://github.com/Azure/application-gateway-kubernetes-ingress/blob/3765a9ad7fc5c2ca808fef8e98a1ef4f26b4919b/pkg/azure/auth.go#L41

And I'm not sure but maybe this whole block may be simplified when using `DefaultAzureCredential` ?

File: helm/ingress-azure/templates/configmap.yaml [hunk 1]
```diff
@@ -78,7 +78,7 @@ data:
 {{- end }}
 
 {{- if .Values.armAuth -}}
-{{- if eq .Values.armAuth.type "aadPodIdentity"}}
+{{- if or (eq .Values.armAuth.type "aadPodIdentity") (eq .Values.armAuth.type "workloadIdentity") }}
```
Comment (jbpaux) [inline helm/ingress-azure/templates/configmap.yaml:81]: do we really need this for workload identity ? `AZURE_CLIENT_ID` is dynamically injected by workload identity webhook mutating controller and not sure what the other variable is used for

Comment (akshaysngupta) [reply helm/ingress-azure/templates/configmap.yaml:81]: `USE_MANAGED_IDENTITY_FOR_POD` is required to bypass the cloud provider config based authorizer.

`AZURE_CLIENT_ID` - yes, this is automatically injected. But I keep it as this doesn't do any harm.

**Review by sean-jeffrey — APPROVED** [review]

**Comment by jkroepke** [comment]
@akshaysngupta can you create a release which bundles this PR?

**Comment by akshaysngupta** [comment]
Hoping to create a release this upcoming week.

**Comment by jbpaux** [comment]
Can you also update the doc https://github.com/Azure/application-gateway-kubernetes-ingress/blob/master/docs/helm-values-documenation.md ?

**Comment by jhoeflaken** [comment]
Is this fix already released?

**Comment by jbpaux** [comment]
In RC : https://github.com/Azure/application-gateway-kubernetes-ingress/blob/master/CHANGELOG/CHANGELOG-1.7.md

**Comment by jhoeflaken** [comment]
Thnx @jbpaux. Got it working!

**Comment by mhauffman** [comment]
> Thnx @jbpaux. Got it working!

@jhoeflaken  I can't seem to get it working on my end. Did you have to do anything besides updating the type to 'workloadIdentity' and using version 1.7.0-rc1?

I seem to keep getting the error below:
`I0221 00:57:04.489425       1 utils.go:114] Using verbosity level 3 from environment variable APPGW_VERBOSITY_LEVEL
I0221 00:57:04.566778       1 supported_apiversion.go:70] server version is: 1.24.9
I0221 00:57:04.583401       1 environment.go:294] KUBERNETES_WATCHNAMESPACE is not set. Watching all available namespaces.
I0221 00:57:04.583429       1 main.go:118] Using User Agent Suffix='agic-ingress-azure-759f9d45bf-fc5tq' when communicating with ARM
I0221 00:57:04.583508       1 main.go:137] Application Gateway Details: Subscription="xx" Resource Group="xx" Name="xxx"
I0221 00:57:04.583515       1 auth.go:58] Creating authorizer using Default Azure Credentials
I0221 00:57:04.583751       1 httpserver.go:57] Starting API Server on :8123
E0221 00:57:04.595137       1 authorizer.go:46] Error getting Azure token: DefaultAzureCredential authentication failed
GET http://169.254.169.254/metadata/identity/oauth2/token
--------------------------------------------------------------------------------
RESPONSE 400 Bad Request
--------------------------------------------------------------------------------
{
  "error": "invalid_request",
  "error_description": "Identity not found"
}
--------------------------------------------------------------------------------
E0221 00:57:04.649217       1 client.go:184] configuration error (bad request) or unauthorized error while performing a GET using the authorizer
E0221 00:57:04.649234       1 client.go:185] stopping GET retries
F0221 00:57:04.649294       1 main.go:175] Failed getting Application Gateway: Code="ErrorApplicationGatewayUnexpectedStatusCode" Message="Unexpected status code '401' while performing a GET on Application Gateway." InnerError="network.ApplicationGatewaysClient#Get: Failure responding to request: StatusCode=401 -- Original Error: autorest/azure: Service returned an error. Status=401 Code="AuthenticationFailedMissingToken" Message="Authentication failed. The 'Authorization' header is missing the access token.""`

And I don't seem to see the environment variables mentioned above being added to the pod.

Appreciate any assistance.

**Comment by AnthonyDewhirst** [comment]
How long does it "normally" take for rc to become ga?

**Comment by jkroepke** [comment]
@AnthonyDewhirst 

it is [GA](https://github.com/Azure/application-gateway-kubernetes-ingress/releases/tag/1.7.0), but it has bugs: https://github.com/Azure/application-gateway-kubernetes-ingress/issues/1532

**Pull request merged**